src/garage/replay_buffer/her_replay_buffer.py

```diff
@@ -2,145 +2,175 @@
+                next_observations are flattened np.ndarrays and not dicts.
+        """
+        transitions = super().sample_transitions(batch_size)
+        obses = []
+        next_obses = []
+        for idx, _ in enumerate(transitions['observations']):
+
+            obses.append(
+                dict(observation=transitions['observations'][idx],
+                     achieved_goal=transitions['achieved_goals'][idx],
+                     desired_goal=transitions['desired_goals'][idx]))
+            next_obses.append(
+                dict(observation=transitions['next_observations'][idx],
+                     achieved_goal=transitions['next_achieved_goals'][idx],
+                     desired_goal=transitions['next_desired_goals'][idx]))
+
+        transitions[
+            'observations'] = self._env_spec.observation_space.flatten_n(obses)
+        transitions[
+            'next_observations'] = self._env_spec.observation_space.flatten_n(
+                next_obses)
+
+        del transitions['achieved_goals']
+        del transitions['desired_goals']
+        del transitions['next_achieved_goals']
+        del transitions['next_desired_goals']
 
-    """
+        return transitions
 
-    def __init__(self, replay_k, reward_fun, env_spec, size_in_transitions,
-                 time_horizon):
-        self._env_spec = env_spec
-        self._sample_transitions = make_her_sample(replay_k, reward_fun)
-        self._replay_k = replay_k
-        self._reward_fun = reward_fun
-        super().__init__(env_spec, size_in_transitions, time_horizon)
+    def add_path(self, path):
+        """Adds a path to the replay buffer.
 
-    def sample(self, batch_size):
-        """Sample a transition of batch_size.
+        For each transition in the given path except the last one,
+        replay_k HER transitions will added to the buffer in addition
+        to the one in the path. The last transition is added without
+        sampling additional HER goals.
 
         Args:
-            batch_size (int): Batch size to sample.
-
-        Return:
-            dict[numpy.ndarray]: Transitions which transitions[key] has the
-                shape of :math:`(N, S^*)`. Keys include [`observation`,
-                `action`, `goal`, `achieved_goal`, `terminal`,
-                `next_observation`, `next_achieved_goal` and `reward`].
+            path(dict[string:np.ndarray]): Each key in the dict must map
+                to a np.ndarray of shape :math:`(T, S^*)`.
 
         """
-        buffer = {}
-        for key in self._buffer:
-            buffer[key] = self._buffer[key][:self._current_size]
-
-        transitions = self._sample_transitions(buffer, batch_size)
-
-        for key in (['reward', 'next_observation', 'next_achieved_goal'] +
-                    list(self._buffer.keys())):
-            assert key in transitions, 'key %s missing from transitions' % key
-
-        return transitions
+        obses = path['observations']
```

Comment:
The logic here to line 133 looks pretty repetitive. Can you somehow decrease the amount of repetition (maybe write a helper that turns a list of observation dictionaries into a dictionary of the flattened observation components)?